launcher/ui/dialogs/BlockedModsDialog.cpp

```diff
@@ -1,28 +1,186 @@
 #include "BlockedModsDialog.h"
-#include "ui_BlockedModsDialog.h"
-#include <QPushButton>
-#include <QDialogButtonBox>
 #include <QDesktopServices>
+#include <QDialogButtonBox>
+#include <QPushButton>
+#include "Application.h"
+#include "ui_BlockedModsDialog.h"
 
+#include <QDebug>
+#include <QStandardPaths>
 
-BlockedModsDialog::BlockedModsDialog(QWidget *parent, const QString &title, const QString &text, const QString &body, const QList<QUrl> &urls) :
-        QDialog(parent), ui(new Ui::BlockedModsDialog), urls(urls) {
+BlockedModsDialog::BlockedModsDialog(QWidget* parent, const QString& title, const QString& text, QList<BlockedMod>& mods)
+    : QDialog(parent), ui(new Ui::BlockedModsDialog), mods(mods)
+{
     ui->setupUi(this);
 
     auto openAllButton = ui->buttonBox->addButton(tr("Open All"), QDialogButtonBox::ActionRole);
     connect(openAllButton, &QPushButton::clicked, this, &BlockedModsDialog::openAll);
 
+    connect(&watcher, &QFileSystemWatcher::directoryChanged, this, &BlockedModsDialog::directoryChanged);
+
+    hashing_task = shared_qobject_ptr<ConcurrentTask>(new ConcurrentTask(this, "MakeHashesTask", 10));
+
+    qDebug() << "Mods List: " << mods;
+
+    setupWatch();
+    scanPaths();
+
     this->setWindowTitle(title);
     ui->label->setText(text);
-    ui->textBrowser->setText(body);
+    ui->labelModsFound->setText(tr("Please download the missing mods."));
+    update();
 }
 
-BlockedModsDialog::~BlockedModsDialog() {
+BlockedModsDialog::~BlockedModsDialog()
+{
     delete ui;
 }
 
-void BlockedModsDialog::openAll() {
-    for(auto &url : urls) {
-        QDesktopServices::openUrl(url);
+void BlockedModsDialog::openAll()
+{
+    for (auto& mod : mods) {
+        QDesktopServices::openUrl(mod.websiteUrl);
+    }
+}
+
+/// @brief update UI with current status of the blocked mod detection
+void BlockedModsDialog::update()
+{
+    QString text;
+    QString span;
+
+    for (auto& mod : mods) {
+        if (mod.matched) {
+            // &#x2714; -> html for HEAVY CHECK MARK : ✔
+            span = QString(tr("<span style=\"color:green\"> &#x2714; Found at %1 </span>")).arg(mod.localPath);
+        } else {
+            // &#x2718; -> html for HEAVY BALLOT X : ✘
+            span = QString(tr("<span style=\"color:red\"> &#x2718; Not Found </span>"));
+        }
+        text += QString(tr("%1: <a href='%2'>%2</a> <p>Hash: %3 %4</p> <br/>")).arg(mod.name, mod.websiteUrl, mod.hash, span);
     }
+
+    ui->textBrowser->setText(text);
+
+    if (allModsMatched()) {
+        ui->labelModsFound->setText(tr("All mods found ✔"));
+    } else {
+        ui->labelModsFound->setText(tr("Please download the missing mods."));
+    }
+}
+
+/// @brief Signal fired when a watched direcotry has changed
+/// @param path the path to the changed directory
+void BlockedModsDialog::directoryChanged(QString path)
+{
+    qDebug() << "Directory changed: " << path;
+    scanPath(path);
+}
```

Comment:
This could cause severe slowdown if the whole directory is being re-scanned every time a new file is added :thinking:

Reply:
There whole directory is not re-scanned, only files that match the right filenames have their hash's checked.
the scan effectively becomes loop over the current filenames. even if there was an attempt to detect which files are "new" since the last check this loop would need to be done. additionally some browsers do not use a part file and download the file directly to the target name and so we have to recheck even files we checked before if their contents change.